### ISSUE TITLE
ack timeout

### ISSUE THREAD

**Issue by yoobi** [spring-projects/spring-amqp#1427]
**Expected Behavior**
Same behaviour as https://www.rabbitmq.com/consumers.html#acknowledgement-timeout

We could have a setter in the RabbitTemplate, or somewhere else

**Current Behavior**
I haven't seen one in the documentation

Below my code to simulate process in receiving:

```kotlin
@RestController
class DemoEventController(private val rabbitTemplate: RabbitTemplate) {

    @PostMapping("/event")
    fun postEventMessage(): ResponseEntity<Void> {
        for(i in 0..5) {
            rabbitTemplate.convertAndSend(RabbitMQConfig.QK_EXAMPLE_QUEUE, "ExampleMessage n°$i")
        }
        return ResponseEntity(HttpStatus.OK)
    }
}


@Component
class DemoEventConsumer(private val template: RabbitTemplate) {
    @RabbitListener(queues = [RabbitMQConfig.QK_EXAMPLE_QUEUE])
    @Throws(IOException::class)
    fun onMessageReceived(message: String, channel: Channel, @Header(AmqpHeaders.DELIVERY_TAG) tag: Long) {
        println("Consumer Message received!: $message at $timeNowRepresentation")
        Thread.sleep(5000L)
        channel.basicAck(tag, false)
        println("Consumer ack message: $message at $timeNowRepresentation")
    }
}

@Component
class DemoEventConsumer2(private val template: RabbitTemplate) {
    @RabbitListener(queues = [RabbitMQConfig.QK_EXAMPLE_QUEUE])
    @Throws(IOException::class)
    fun onMessageReceived(message: String, channel: Channel, @Header(AmqpHeaders.DELIVERY_TAG) tag: Long) {
        println("Consumer2 Message received!: $message at $timeNowRepresentation")
       Thread.sleep(5000L)
        channel.basicAck(tag, false)
        println("Consumer2 ack message: $message at $timeNowRepresentation")
    }
}

private val timeNowRepresentation: String
    get() {
        val now = System.currentTimeMillis()
        val simpleDateFormat = SimpleDateFormat("HH:mm:ss")
        return simpleDateFormat.format(now)
    }
```

**Context**
I have one queue and two consumers, and I want that if a consumer takes a long time to process it is automatically nack from it and put back in the queue to be processed again until he is ack.

**Comment by garyrussell** [comment]
It is not clear what you mean; if the broker times out the consumer, the channel is closed and the message redelivered.

There is no way to simulate this behavior on the consumer side; the only thing we could do is interrupt the consumer thread when the broker closes the channel (which would break it out of its sleep) but that won't solve every situation - e.g. blocked on a non-interruptible operation.

**Comment by yoobi** [comment]
Hello, sorry if it's not clear, I'm currently learning RabbitMQ. I'll try to reformulate with an example.
- 1 Queue with 3 messages in round-robin.
- 2 Consumers (C1 & C2)
- To complete the task related to the message it takes 3 seconds

Workflow:
1. Queue has M1, M2 and M3 waiting to be processed
2. C1 gets M1 and completes it in 2 seconds, C1 sends ack to Queue so it can delete the M1 from it. (Queue has M2 and M3)
3. C2 gets M2 and completes it in 2 seconds, C2 sends ack to Queue so it can delete the M2 from it (Queue has M3)
4. C1 gets M3 and after 3 seconds the task is still not complete, C1 should sends nack to Queue and stop the processing of M3. (Queue still has M3)
5. C1 gets M3 again as it is not processing anything. M3 is completed under 3 seconds, C1 sends ack to Queue to delete M3 from it. (Queue is empty)

In other words, if the acknowledgment isn't done in the period specified it is nack from the consumer and another consumer can pick it up.

**Comment by garyrussell** [comment]
That's not possible; the container has no control over the channel while the listener is processing.

You could add your own task to monitor how long the listener is running for and call `reset()` on the connection factory to force the message to be requeued, but your listener would continue to process and would get an error when trying to ack the message. You would need some way to stop the listener from completing its work.
